src/Microsoft.OData.Core/ODataReaderCore.cs

```diff
@@ -946,6 +946,33 @@ private T InterceptException<T>(Func<T> action)
             }
         }
 
+        /// <summary>
+        /// Catch any exception thrown by the action passed in; in the exception case move the reader into
+        /// state ExceptionThrown and then rethrow the exception.
+        /// </summary>
+        /// <typeparam name="T">The type returned from the <paramref name="action"/> to execute.</typeparam>
+        /// <param name="action">The action to execute.</param>
+        /// <returns>
+        /// A task that represents the asynchronous operation.
+        /// The value of the TResult parameter contains the result of executing the <paramref name="action"/>.
+        /// </returns>
+        private async Task<T> InterceptExceptionAsync<T>(Func<ODataReaderCore, Task<T>> action)
+        {
+            try
+            {
+                return await action(this).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                if (ExceptionUtils.IsCatchableExceptionType(e))

```

Comment:
I'm sure we don't care very much about optimizing the failure cases, but catching all exceptions to only then figure out if we can handle them makes exceptions that aren't expected more expensive

Reply:
@corranrogue9 What we do in `IsCatchableExceptionType` method is to check that we're not dealing with an `OutOfMemoryException`. We do a little optimization in that method by storing the `OutOfMemoryException` type in a static field